File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/offline/KagemushaRecursiveAggregationProofBundleProver.java
package org.hyperledger.iroha.android.offline;

/** Native record-backed Kagemusha recursive aggregation proof-bundle prover. */
public final class KagemushaRecursiveAggregationProofBundleProver {
  private static final String LIBRARY_NAME = "connect_norito_bridge";
  private static final boolean NATIVE_AVAILABLE = loadLibrary();

  private KagemushaRecursiveAggregationProofBundleProver() {}

  public static boolean isNativeAvailable() {
    return NATIVE_AVAILABLE;
  }

  public static byte[] proveVerifiedRecursiveAggregationProofBundleWithRecordsAndPallasOpenEnvelopes(
      final byte[] recordBundleArchive, final byte[] pallasOpenEnvelopesArchive) {
    if (recordBundleArchive == null || recordBundleArchive.length == 0) {
      throw new IllegalArgumentException("recordBundleArchive must not be empty");
    }
    if (pallasOpenEnvelopesArchive == null || pallasOpenEnvelopesArchive.length == 0) {
      throw new IllegalArgumentException("pallasOpenEnvelopesArchive must not be empty");
    }
    requireNative();
    final byte[] proofBundleArchive =
        nativeProveVerifiedRecursiveAggregationProofBundleWithRecordsAndPallasOpenEnvelopes(
            recordBundleArchive, pallasOpenEnvelopesArchive);
    return KagemushaCompactPaymentTokenProver.requireNativeOutput(
        proofBundleArchive,
        "nativeProveVerifiedRecursiveAggregationProofBundleWithRecordsAndPallasOpenEnvelopes");
  }

  private static void requireNative() {
    if (!NATIVE_AVAILABLE) {
      throw new IllegalStateException(LIBRARY_NAME + " is not available in this runtime");
    }
  }

  private static boolean loadLibrary() {
    return detectNativeAvailability(
        () -> System.loadLibrary(LIBRARY_NAME),
        () ->
            expectIllegalArgumentProbe(
                () ->
                    nativeProveVerifiedRecursiveAggregationProofBundleWithRecordsAndPallasOpenEnvelopes(
                        new byte[0], new byte[0])));
  }

  static boolean detectNativeAvailability(
      final NativeProbe loadLibrary, final NativeSymbolProbe probeSymbol) {
    try {
      loadLibrary.run();
    } catch (final IllegalArgumentException error) {
      return false;
    } catch (final UnsatisfiedLinkError | SecurityException error) {
      return false;
    }
    try {
      return probeSymbol.run();
    } catch (final IllegalArgumentException error) {
      return false;
    } catch (final UnsatisfiedLinkError | SecurityException error) {
      return false;
    }
  }

  interface NativeProbe {
    void run();
  }

  interface NativeSymbolProbe {
    boolean run();
  }

  static boolean expectIllegalArgumentProbe(final NativeProbe probe) {
    try {
      probe.run();
      return false;
    } catch (final IllegalArgumentException expected) {
      return true;
    }
  }

  private static native byte[]
      nativeProveVerifiedRecursiveAggregationProofBundleWithRecordsAndPallasOpenEnvelopes(
          byte[] recordBundleArchive, byte[] pallasOpenEnvelopesArchive);
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/offline/KagemushaRecursiveAggregationProofBundleProver.kt
package org.hyperledger.iroha.sdk.offline

/** Native record-backed Kagemusha recursive aggregation proof-bundle prover. */
class KagemushaRecursiveAggregationProofBundleProver private constructor() {
    companion object {
        private const val LIBRARY_NAME = "connect_norito_bridge"
        private val nativeAvailable: Boolean = loadLibrary()

        @JvmStatic
        fun isNativeAvailable(): Boolean = nativeAvailable

        @JvmStatic
        fun proveVerifiedRecursiveAggregationProofBundleWithRecordsAndPallasOpenEnvelopes(
            recordBundleArchive: ByteArray?,
            pallasOpenEnvelopesArchive: ByteArray?,
        ): ByteArray {
            KagemushaCompactPaymentTokenProver.requireNativeInput(
                recordBundleArchive,
                "recordBundleArchive",
            )
            KagemushaCompactPaymentTokenProver.requireNativeInput(
                pallasOpenEnvelopesArchive,
                "pallasOpenEnvelopesArchive",
            )
            val recordBundle = KagemushaCompactPaymentTokenProver.ownedNativeInput(
                recordBundleArchive,
                "recordBundleArchive",
            )
            val pallasOpenEnvelopes = KagemushaCompactPaymentTokenProver.ownedNativeInput(
                pallasOpenEnvelopesArchive,
                "pallasOpenEnvelopesArchive",
            )
            check(nativeAvailable) { "$LIBRARY_NAME is not available in this runtime" }
            val proofBundleArchive =
                nativeProveVerifiedRecursiveAggregationProofBundleWithRecordsAndPallasOpenEnvelopes(
                    recordBundle,
                    pallasOpenEnvelopes,
                )
            return KagemushaCompactPaymentTokenProver.requireNativeOutput(
                proofBundleArchive,
                "nativeProveVerifiedRecursiveAggregationProofBundleWithRecordsAndPallasOpenEnvelopes",
            )
        }

        private fun loadLibrary(): Boolean =
            detectNativeAvailability(
                loadLibrary = { System.loadLibrary(LIBRARY_NAME) },
                probeSymbol = {
                    expectIllegalArgumentProbe {
                        nativeProveVerifiedRecursiveAggregationProofBundleWithRecordsAndPallasOpenEnvelopes(
                            ByteArray(0),
                            ByteArray(0),
                        )
                    }
                },
            )

        internal fun detectNativeAvailability(loadLibrary: () -> Unit, probeSymbol: () -> Boolean): Boolean {
            try {
                loadLibrary()
            } catch (_: IllegalArgumentException) {
                return false
            } catch (_: UnsatisfiedLinkError) {
                return false
            } catch (_: SecurityException) {
                return false
            } catch (_: RuntimeException) {
                return false
            }
            return try {
                probeSymbol()
            } catch (_: IllegalArgumentException) {
                false
            } catch (_: UnsatisfiedLinkError) {
                false
            } catch (_: SecurityException) {
                false
            } catch (_: RuntimeException) {
                false
            }
        }

        internal fun expectIllegalArgumentProbe(probe: () -> Unit): Boolean =
            try {
                probe()
                false
            } catch (_: IllegalArgumentException) {
                true
            }

        @JvmStatic
        private external fun nativeProveVerifiedRecursiveAggregationProofBundleWithRecordsAndPallasOpenEnvelopes(
            recordBundleArchive: ByteArray,
            pallasOpenEnvelopesArchive: ByteArray,
        ): ByteArray?
    }
}
